### ISSUE TITLE
Project dependencies have API risk issues

### ISSUE THREAD

**Issue by PyDR-Tool** [cleardusk/3DDFA#218]
Hi, In **3DDFA**, inappropriate dependency versioning constraints can cause risks.

Below are the dependencies and version constraints that the project is using
```
torch>=0.4.1
torchvision>=0.2.1
numpy>=1.15.4
scipy>=1.1.0
matplotlib==3.0.2
dlib==19.5.0
opencv-python>=3.4.3.18
```

The version constraint **==** will introduce the risk of dependency conflicts because the scope of dependencies is too strict.
The version constraint **No Upper Bound** and **\*** will introduce the risk of the missing API Error because the latest version of the dependencies may remove some APIs.

After further analysis, in this project,
The version constraint of dependency **scipy** can be changed to *>=0.8.0,<=1.2.3*.

The above modification suggestions can reduce the dependency conflicts as much as possible, 
and introduce the latest version as much as possible without calling Error in the projects.

The invocation of the current project includes all the following methods.

    
<details>
<summary>The calling methods from the scipy</summary>
<pre>
io.imread
</pre>
</details>

 

    
<details>
<summary>The calling methods from the  all methods</summary>
<pre>
img_step2.transform.unsqueeze
_predict_vertices
crop_img
gen_img_paf
data_time.update
point_2d.reshape
cv2.imshow
input_.data.clone
rect_fp.open.read
torch.abs
imageio.imread
self.dw5_6
np.float32.BG.astype.copy
plt.figure
int
cget_depths_image
_to_tensor
pickle.load
rects.append
min
ax.axis
ncc
rect.top
np.abs
np.zeros
torch.randperm
logging.info
k.replace
asin
plt.savefig
logging.StreamHandler
self.dw2_1
np.int32
dlib.rectangle
index.index.index.torch.cat.view
cpncc
data.DataLoader
checkpoint.keys
obama_demo
build_camera_box
adjust_learning_rate
alpha_expg.self.w_exp.alpha_shpg.self.w_shp.self.u.view
self.reset
torch.no_grad
param.squeeze.cpu.numpy.flatten.astype
pickle.dump
RenderPipeline
cv2.line
_tensor_to_cuda
np.minimum
model
aflw
m.weight.data.normal_
tuple
gen_offsets
self.conv1
np.savetxt
args.milestones.split
np.dot
device_ids.model.nn.DataParallel.cuda
p_.reshape
torch.save
transform
param.view
torch.randn
_parse_param_batch
dlib.get_frontal_face_detector
p_.view
ana_aflw
x.size
batch_time.update
input.cuda
plot_close
self.VDCLoss.super.__init__
self.mean.tensor.sub_.div_
header.format
io.imread
img.transform.unsqueeze
torch.norm
torch.optim.SGD
mesh_core_cython.get_normal
vertices_lst.append
self.conv_sep
calc_nme_alfw
self._calc_weights_resample
np.float32.colors.astype.copy
plt.imshow
rect.left
v.lower
calc_nme_alfw2000
benchmark_alfw_params
cv2.imread
ax.plot3D
mesh_core_cython.render_colors_core
point_3d_homo.dot
self.forward_all
depths_img.squeeze
time.time
nn.Conv2d
WPDCLoss
_load
self.MobileNet.super.__init__
self.bn_sep
args.opt_style.WPDCLoss.cuda
logging.FileHandler
np.array
losses.update
face_regressor
type
nn.DataParallel
param.squeeze.cpu
np.zeros_like
draw_landmarks
aflw2000
self.WPDCLoss.super.__init__
last_frame_pts.append
nn.AdaptiveAvgPool2d
self.dw6
alpha_expg.w_exp_base.alpha_shpg.w_shp_base.u_base.view
NormalizeGjz
maxes.view
dlib.shape_predictor
rect.frame.face_regressor.parts
face_detector
super
ax.set_xticklabels
save_checkpoint
gen_3d_vertex
transforms.Compose
np.linalg.norm
loss.item
_to_ctype
enumerate
torch.zeros_like
plt.tight_layout
osp.join
cos
weights.max
vertices.np.max.reshape
plt.figaspect
ax.set_zticklabels
max
tensor.sub_
sys.path.append
app
atan2
parser.parse_args
model.cuda
outputs.append
nme_list.append
render_img.astype
sio.savemat
torch.tensor
vars
get_colors
DDFATestDataset
np.float.point_3d.np.array.reshape
sio.loadmat
ax.imshow
model.parameters
N.alpha_exp.w_exp_base.alpha_shp.w_shp_base.u_base.view.permute
param.reshape
self.resample_num.self.w_shp_length.torch.randperm.reshape
img_fp.imageio.imread.astype
colors.astype
loss.mean
pnccs_img.squeeze
inputs.cuda
imgs.append
calc_hypotenuse
arr.copy
self.dw5_3
kwargs.get
np.int32.triangles.astype.copy
self.dw3_2
convert_type
filelists.Path.read_text
cv2.waitKey
image.copy
np.min
os.system
test
_benchmark_aflw
imageio.mimwrite
filelists.open.read
ToTensorGjz
np.max
setup
vc.read
meta.get
mode.lower
dump_to_ply
plt.subplots_adjust
alpha_exp.w_exp_base.alpha_shp.w_shp_base.u_base.reshape
ax.set_yticklabels
np.ones
self.reconstruct_and_parse
reconstruct_paf_anchor
np.concatenate
sorted
model.train
obj_name.split
np.where
filelists.Path.read_text.strip.split
self.bn1
nn.Linear
range
cv2.polylines
model.eval
P2sRt
param.squeeze.cpu.numpy.flatten
imageio.imwrite
MobileNet
self.fc
i.output.cpu
parse_pose
ax.plot
argparse.ArgumentParser
np.sum
cv2.resize
img.float
x.cuda
print_args
os.mkdir
filelists.open.read.strip.split
nn.MSELoss
_numpy_to_tensor
self.transform
predict_68pts
alpha_exp.self.w_exp.alpha_shp.self.w_shp.self.u.view
open
filelists.open.read.strip
rect_fp.open.read.strip
pts_res.append
parse_args
render.crender_colors
torch.load
x.cpu.numpy
np.mean
osp.dirname
_benchmark_aflw2000
t3d.reshape
l.split
self.forward_resample
np.float32.vertices.astype.copy
timeit.repeat
norm_vertices
self.dw5_4
np.maximum
resample_num.self.w_shp_length.torch.randperm.reshape
filename.rfind
arch.mobilenet_v1.getattr
self.avgpool
param.squeeze
self.conv_dw
np.floor
self.dw4_2
main
i.output.cpu.numpy.flatten
rect.img_ori.face_regressor.parts
kpt.np.round.astype
imageio.imread.astype
is_point_in_tri
map
alpha_exp.w_exp_base.alpha_shp.w_shp_base.u_base.view
cv2.imwrite
losses.append
img_fp.replace
np.std
_load_cpu
criterion
numpy.get_include
dump_vertex
Extension
plot_pose_box
vertices.min
crender_colors
self.dw3_1
rect.bottom
index0.index1.img_crop.reshape.transpose
get_suffix
print
cv2.VideoCapture
glob
np.round
_dump
poses.append
model.load_state_dict
paf.get
N.alpha_expg.w_exp_base.alpha_shpg.w_shp_base.u_base.view.permute
_get_suffix
np.bitwise_and
to
os.path.exists
alpha_exp.w_exp_filter.alpha_shp.w_filter.u_filter.reshape
alpha_exp.w_exp.alpha_shp.w_shp.u.reshape
model_dict.keys
getattr
args.devices_id.model.nn.DataParallel.cuda
args.loss.lower
block
torch.from_numpy
Path
np.ceil
np.load
list
self._target_loader
N.alpha_expg.self.w_exp.alpha_shpg.self.w_shp.self.u.view.permute
predict_dense
osp.split
self.dw5_5
model.astype
extract_param
plt.plot
mkdir
N.alpha_exp.self.w_exp.alpha_shp.self.w_shp.self.u.view.permute
plt.close
benchmark_aflw2000_params
ana_alfw2000
join
triangles.copy
index0.index1.img_crop.reshape
nn.PReLU
triangles._to_ctype.astype
VDCLoss
f.write
ax.scatter
np.cross
triangles.astype
parse_roi_box_from_landmark
sqrt
anchor.np.round.astype
keypoints.T.flatten
BG.astype
rect_fp.open.read.strip.split
filelists.Path.read_text.strip
input.size
keypoints.u.reshape
_norm
self.bn_dw
train
args.size_average.nn.MSELoss.cuda
args.devices_id.split
optimizer.zero_grad
loss.backward
osp.basename
model_path.replace
x.view
reconstruct_vertex
args.opt_style.VDCLoss.cuda
args.arch.mobilenet_v1.getattr
fn.replace
parser.add_argument
write_obj_with_colors
DDFADataset
kc.replace
vertices.max
torch.mean
os.path.isdir
cv2.circle
vertices.copy
rect.right
optimizer.step
validate
fig.add_subplot
img_loader
Ps.append
m.bias.data.zero_
target.cuda
np.hstack
torch.onnx.export
self.dw5_1
format
param.squeeze.cpu.numpy
nn.ReLU
isinstance
Exception
self.dw5_2
_parse_param
osp.realpath
convert_to_ori
self.dw2_2
torch.cat
str
fp._load.torch.from_numpy.cuda
logging.basicConfig
m.weight.data.fill_
self.DepthWiseBlock.super.__init__
render_colors
make_abs_path
vertices.astype
math.sqrt
self.img_loader
AverageMeter
np.sqrt
self.relu
vertices.np.min.reshape
len
np.save
np.clip
DataLoader
benchmark_pipeline
plt.axis
x.cpu
argparse.ArgumentTypeError
checkpoint_fp.replace
matrix2angle
plt.show
vertices.np.round.astype
np.power
model.state_dict
x.split.replace
x.split
i.output.cpu.numpy
target_.data.clone
round
pic.transpose
u_exp.u_shp.astype
parse_roi_box_from_bbox
index.index.index.torch.cat.view.cuda
dlib.rectangles
point_3d.append
self.dw4_1
osp.exists
ax.view_init
nn.BatchNorm2d
self.modules
args.resume.Path.is_file
torch.cuda.set_device
</pre>
</details>

Could please help me check this issue?
May I pull a request to fix it?
Thank you very much.
